Lib/fontbakery/specifications/ufo_sources.py

```diff
@@ -0,0 +1,205 @@
+# {based_on_style: google}'` and `docformatter`.
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import os
+
+from fontbakery.callable import check, condition
+from fontbakery.checkrunner import ERROR, FAIL, PASS, WARN, Section, Spec
+from fontbakery.constants import CRITICAL
+
+
+class UFOSpec(Spec):
+
+    def setup_argparse(self, argument_parser):
+        """Set up custom arguments needed for this spec."""
+        import glob
+        import logging
+        import argparse
+
+        def get_fonts(pattern):
+
+            fonts_to_check = []
+            # use glob.glob to accept *.ufo
+
+            for fullpath in glob.glob(pattern):
+                fullpath_absolute = os.path.abspath(fullpath)
+                if fullpath_absolute.endswith(".ufo") and os.path.isdir(
+                        fullpath_absolute):
+                    fonts_to_check.append(fullpath)
+                else:
+                    logging.warning(
+                        ("Skipping '{}' as it does not seem "
+                         "to be valid UFO source directory.").format(fullpath))
+            return fonts_to_check
+
+        class MergeAction(argparse.Action):
+
+            def __call__(self, parser, namespace, values, option_string=None):
+                target = [item for l in values for item in l]
+                setattr(namespace, self.dest, target)
+
+        argument_parser.add_argument(
+            'fonts',
+            nargs='+',
+            type=get_fonts,
+            action=MergeAction,
+            help='font file path(s) to check.'
+            ' Wildcards like *.ufo are allowed.')
+        return ('fonts',)
+
+
+# ----------------------------------------------------------------------------
+# This variable serves as an exportable anchor point, see e.g. the
+# Lib/fontbakery/commands/check_ufo_sources.py script.
+specification = UFOSpec(
+    default_section=Section('Default'),
+    iterargs={'font': 'fonts'},
+    derived_iterables={'ufo_sources': ('ufo_source', True)})
+
+register_check = specification.register_check
+register_condition = specification.register_condition
+# ----------------------------------------------------------------------------
+
+
+@register_condition
+@condition
+def ufo_source(font):
+    import defcon
+    return defcon.Font(font)
+
+
+@register_check
+@check(id='com.daltonmaag/check/ufolint', priority=CRITICAL)
+def com_daltonmaag_check_ufolint(font):
+    """Run ufolint on UFO source directory."""
+    import subprocess
+    ufolint_cmd = ["ufolint", font]
+
+    try:
+        subprocess.check_output(ufolint_cmd, stderr=subprocess.STDOUT)
+    except subprocess.CalledProcessError as e:
+        yield FAIL, ("ufolint failed the UFO source. Output follows :"
+                     "\n\n{}\n").format(e.output)
+    except OSError:
+        yield ERROR, "ufolint is not available!"
+
+    yield PASS, "ufolint passed the UFO source."
```

Comment:
you need a boolean initialized as failed=False then make it True in the exception handlers above and then do "if not failed yield pass" here.

Reply:
that's a general pattern used throughout our codebase, actually

Reply:
Oh, why not use `else:` appended to the try block?

Reply:
the `else` branch runs when there is no exception, before the `finally` clause. That is its primary purpose. Here there is no `finally` so it doesn't make any difference

Reply:
Another use of `else` is to do something only when no exception occurs and to _not_ do that when exceptions are handled. In this case there is a `yield` in the `except:` branches, not a `return`, so technically Nikolaus is correct: if one where to call `next()` on the iterator returned by this function after a `FAIL` or an `ERROR`, it will surprisingly continue execution and meet a `yield PASS`. So yes, it would make sense to protect it with an `else`